File: docs/code-examples/arrow3d_simple.rs
//! Log a batch of 3D arrows.

use std::f32::consts::TAU;

use rerun::{
    archetypes::Arrows3D,
    components::{Color, Vector3D},
    RecordingStreamBuilder,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (rec, storage) = RecordingStreamBuilder::new("rerun_example_arrow3d").memory()?;

    let (vectors, colors): (Vec<_>, Vec<_>) = (0..100)
        .map(|i| {
            let angle = TAU * i as f32 * 0.01;
            let length = ((i + 1) as f32).log2();
            let c = (angle / TAU * 255.0).round() as u8;
            (
                Vector3D::from([(length * angle.sin()), 0.0, (length * angle.cos())]),
                Color::from_unmultiplied_rgba(255 - c, c, 128, 128),
            )
        })
        .unzip();

    rec.log("arrows", &Arrows3D::new(vectors).with_colors(colors))?;

    rerun::native_viewer::show(storage.take())?;
    Ok(())
}

File: docs/code-examples/arrow3d_simple.cpp
// Log a batch of 3D arrows.

#include <rerun.hpp>

#include <cmath>
#include <vector>

constexpr float TAU = 6.28318530717958647692528676655900577f;

int main() {
    const auto rec = rerun::RecordingStream("rerun_example_arrow3d");
    rec.spawn().exit_on_failure();

    std::vector<rerun::Position3D> origins;
    std::vector<rerun::Vector3D> vectors;
    std::vector<rerun::Color> colors;

    for (int i = 0; i < 100; ++i) {
        origins.push_back({0, 0, 0});

        float angle = TAU * static_cast<float>(i) * 0.01f;
        float length = log2f(static_cast<float>(i + 1));
        vectors.push_back({length * sinf(angle), 0.0, length * cosf(angle)});

        uint8_t c = static_cast<uint8_t>(round(angle / TAU * 255.0f));
        colors.push_back({static_cast<uint8_t>(255 - c), c, 128, 128});
    }

    rec.log(
        "arrows",
        rerun::Arrows3D::from_vectors(vectors).with_origins(origins).with_colors(colors)
    );
}
